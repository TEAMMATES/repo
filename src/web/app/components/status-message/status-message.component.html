<div class="alert alert-{{ statusMessage.color }}" *ngFor="let statusMessage of messages">
  <button class="close" (click)="dismiss(statusMessage)">
    <span aria-hidden="true">&times;</span>
  </button>

  <!--TODO: Change this behavior when no status messages are returned in raw HTML-->
  <div style="padding-right: 30px;" [innerHtml]="statusMessage.message"></div>
</div>
