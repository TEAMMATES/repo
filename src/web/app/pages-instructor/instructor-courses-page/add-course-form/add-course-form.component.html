<div class="card border-primary fill-plain">
  <div class="close-header">
    <button type="button" class="close" aria-label="Close" (click)="closeEditFormHandler()">
      <span aria-hidden="true">&times;</span>
    </button>
  </div>
  <div class="card-body">
    <form #courseForm="ngForm" (ngSubmit)="onSubmit()">
      <div class="form-group row">
        <label class="col-sm-3 control-label bold-label text-right">Course ID:</label>
        <div class="col-sm-3">
          <input class="form-control" #courseId="ngModel" type="text" [(ngModel)]="newCourseId" name="newCourseId"
              ngbTooltip="Enter the identifier of the course, e.g.CS3215-2013Semester1." required
              [maxlength]="FormValidator.COURSE_ID_MAX_LENGTH" placeholder="e.g. CS3215-2013Semester1"/>
          <div [hidden]="courseId.valid || (courseId.pristine && courseId.untouched)"
               class="invalid-field">
            <i class="fa fa-exclamation-circle" aria-hidden="true"></i>
            The field Course ID should not be empty.
          </div>
        </div>
      </div>
      <div class="form-group row">
        <label class="col-sm-3 control-label bold-label text-right">Course Name:</label>
        <div class="col-sm-9">
          <input class="form-control" #courseName="ngModel" type="text" [(ngModel)]="newCourseName" name="newCourseName"
              ngbTooltip="Enter the name of the course, e.g. Software Engineering." required
              [maxlength]="FormValidator.COURSE_NAME_MAX_LENGTH" placeholder="e.g. Software Engineering"/>
          <div [hidden]="courseName.valid || (courseName.pristine && courseName.untouched)"
               class="invalid-field">
            <i class="fa fa-exclamation-circle" aria-hidden="true"></i>
            The field Course Name should not be empty.
          </div>
        </div>
      </div>
      <div class="form-group row">
        <label class="col-sm-3 control-label bold-label text-right">Time Zone:</label>
        <div class="col-sm-9">
          <div class="input-group">
            <select class="form-control" #timeZone="ngModel" [(ngModel)]="timezone" name="timezone" required
                    ngbTooltip="The time zone for the course. You should not need to change this as it is auto-detected based on your device settings.
                    TEAMMATES automatically adjusts to match the current time offset in your area, including clock changes due to daylight saving time.">
              <option *ngFor="let timezone of timezones" [value]="timezone">{{ timezone }}</option>
            </select>
            <span class="input-group-btn">
              <button class="btn btn-primary" (click)="onAutoDetectTimezone()">Auto Detect</button>
            </span>
          </div>
        </div>
      </div>
      <div class="form-group row">
        <div class="offset-sm-3 col-sm-9">
          <button class="btn btn-primary" type="submit">Add Course</button>
        </div>
      </div>
    </form>
  </div>
</div>

<ng-template #newCourseMessageTemplate>
  The course has been added. Click <a href="/web/instructor/courses/enroll?courseid={{course.courseId}}">here</a>
  to add students to the course or click <a href="/web/instructor/courses/edit?courseid={{course.courseId}}">here</a>
  to add other instructors.<br>If you don't see the course in the list below,
  please refresh the page after a few moments.
</ng-template>
