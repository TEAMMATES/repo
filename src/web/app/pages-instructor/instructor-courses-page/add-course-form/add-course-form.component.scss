.bold-label {
  font-weight: 700 !important;
}

.fill-plain {
  background-color: #EAEFF5;
}

.close-header {
  padding-top: .75rem;
  padding-right: 1.25rem;
  margin-bottom: 0;
}

.form-group {
  min-width: 700px;
}
