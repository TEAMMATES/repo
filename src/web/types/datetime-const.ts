export enum HoursConst {
    ZERO_HOURS = 0,
    TWELVE_HOURS = 12,
    ONE_DAY_HOURS = 24,
    THREE_DAYS_HOURS = 72,
    ONE_WEEK_HOURS = 168,
}

export enum SecondsConst {
    ONE_MINUTE_SECONDS = 60,
}

export enum MinutesConst {
    ONE_HOURS_MINUTES = 60,
}

export enum MsConst {
    ONE_SECOND_MILLISECONDS = 1000,
    TEN_SECOND_MILLISECONDS = 10000,
    ONE_MINUTE_MILLISECONDS = 60000,
    TEN_MINUTE_MILLISECONDS = 600000,
    FIFTEEN_MINUTE_MILLISECONDS = 900000,
    ONE_HOUR_MILLISECONDS = 3600000,
    ONE_WEEK_MILLISECONDS = 604800000,
}

export const getDefaultDateFormat = (): DateFormat => {
    return { year: 0, month: 0, day: 0 };
};

export const getLatestTimeFormat = (): TimeFormat => {
    return { hour: 23, minute: 59 };
};

export const getDefaultTimeFormat = (): TimeFormat => {
    return { hour: 0, minute: 0 };
};

/**
 * The date format.
 */
export interface DateFormat {
    year: number;
    month: number;
    day: number;
}

/**
 * The output format of the time picker.
 */
export interface TimeFormat {
    hour: number;
    minute: number;
}
