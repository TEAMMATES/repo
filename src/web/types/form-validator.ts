import { ApiConst, ApiStringConst } from './api-const';

/**
 * Represents the root FormValidator object of all form fields.
 */
export enum FormValidator {
  /**
   * Max length for the 'Course ID' field.
   */
  COURSE_ID_MAX_LENGTH = ApiConst.COURSE_ID_MAX_LENGTH,

  /**
   * Max length for the 'Course Name' field.
   */
  COURSE_NAME_MAX_LENGTH = ApiConst.COURSE_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Student Name` field.
   */
  STUDENT_NAME_MAX_LENGTH = ApiConst.STUDENT_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Section Name` field.
   */
  SECTION_NAME_MAX_LENGTH = ApiConst.SECTION_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Team Name` field.
   */
  TEAM_NAME_MAX_LENGTH = ApiConst.TEAM_NAME_MAX_LENGTH,

  /**
   * Max length for the 'E-mail Address` field.
   */
  EMAIL_MAX_LENGTH = ApiConst.EMAIL_MAX_LENGTH,

  /**
   * Regex used to verify URLs as Angular does not have a built-in URL validator.
   */
  URL_REGEX = ApiStringConst.URL_REGEX,

  /**
   * Regex used to verify emails in the back-end.
   */
  EMAIL_REGEX = ApiStringConst.EMAIL_REGEX,
}
