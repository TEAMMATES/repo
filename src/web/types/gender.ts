/**
 * Represents the gender of a student.
 */
export enum Gender {
  MALE,
  FEMALE,
  OTHER,
}
export namespace Gender {

  export const values: () => string[] = (): string[] =>
      Object.keys(Gender).filter((key: string) => isNaN(key as any) && key !== 'values');

  export const enums: () => Gender[] = (): Gender[] => {
    const genders: Gender[] = [];
    genders.push(Gender.MALE);
    genders.push(Gender.FEMALE);
    genders.push(Gender.OTHER);
    return genders;
  };
}
