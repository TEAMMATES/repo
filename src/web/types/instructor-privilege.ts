import { InstructorPrivilege } from './api-output';

/**
 * The default instructor privilege.
 */
export const DEFAULT_INSTRUCTOR_PRIVILEGE: InstructorPrivilege = {
  canEditCourse: false,
  canDeleteCourse: false,
  canRestoreCourse: false,
  canModifySession: false,
  canModifyStudent: false,
  canSubmitSessionInSections: false,
  canModifyInstructor: false,
  canViewStudentInSections: false,
  canModifySessionCommentsInSections: false,
  canViewSessionInSections: false,
};
