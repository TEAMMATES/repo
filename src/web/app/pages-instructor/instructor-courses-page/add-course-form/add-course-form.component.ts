import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import moment from 'moment-timezone';
import { HttpRequestService } from '../../../../services/http-request.service';
import { StatusMessageService } from '../../../../services/status-message.service';
import { TimezoneService } from '../../../../services/timezone.service';
import { MessageOutput } from '../../../../types/api-output';
import { ErrorMessageOutput } from '../../../error-message-output';

/**
 * The actual component
 */
@Component({
  selector: 'tm-add-course-form',
  templateUrl: './add-course-form.component.html',
  styleUrls: ['./add-course-form.component.scss'],
})
export class AddCourseFormComponent implements OnInit {

  user: string = '';

  @Input() isEnabled: boolean = true;
  @Output() courseAdded: EventEmitter<void> = new EventEmitter<void>();

  timezones: string[] = [];
  timezone: string = '';
  newCourseId: string = '';
  newCourseName: string = '';

  constructor(private route: ActivatedRoute,
              private httpRequestService: HttpRequestService,
              private statusMessageService: StatusMessageService,
              private timezoneService: TimezoneService) { }

  ngOnInit(): void {
    if (!this.isEnabled) {
      this.timezones = ['UTC', 'Other options ommitted...'];
      this.timezone = 'UTC';
      return;
    }
    this.route.queryParams.subscribe((queryParams: any) => {
      this.user = queryParams.user;
    });
    this.timezones = Object.keys(this.timezoneService.getTzOffsets());
    this.timezone = moment.tz.guess();
  }

  /**
   * Auto-detects timezone for instructor.
   */
  onAutoDetectTimezone(): void {
    if (!this.isEnabled) {
      return;
    }
    this.timezone = moment.tz.guess();
  }

  /**
   * Submits the data to add the new course.
   */
  onSubmit(): void {
    if (!this.isEnabled) {
      return;
    }
    if (!this.newCourseId || !this.newCourseName) {
      this.statusMessageService.showErrorMessage(
          'Please make sure you have filled in both Course ID and Name before adding the course!');
      return;
    }
    const paramMap: { [key: string]: string } = {
      courseid: this.newCourseId,
      coursename: this.newCourseName,
      coursetimezone: this.timezone,
      user: this.user,
    };
    this.newCourseId = '';
    this.newCourseName = '';
    this.timezone = moment.tz.guess();
    this.httpRequestService.post('/instructor/courses', paramMap).subscribe((resp: MessageOutput) => {
      this.courseAdded.emit();
      this.statusMessageService.showSuccessMessage(resp.message);
    }, (resp: ErrorMessageOutput) => {
      this.statusMessageService.showErrorMessage(resp.error.message);
    });
  }

}
