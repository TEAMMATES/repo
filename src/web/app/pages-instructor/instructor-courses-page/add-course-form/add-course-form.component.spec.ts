import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { CourseService } from '../../../../services/course.service';
import { StatusMessageService } from '../../../../services/status-message.service';
import { TimezoneService } from '../../../../services/timezone.service';
import { Course } from '../../../../types/api-output';

import { HttpClientTestingModule } from '@angular/common/http/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MatSnackBarModule } from '@angular/material';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';

import { of } from 'rxjs';
import { AddCourseFormComponent } from './add-course-form.component';

describe('AddCourseFormComponent', () => {
  let component: AddCourseFormComponent;
  let fixture: ComponentFixture<AddCourseFormComponent>;

  const date1: Date = new Date('2020-13-05T08:15:30');
  const courseId1: string = 'CS3281';
  const courseName1: string = 'Valid course';
  const timeZone1: string = 'UTC';
  const course1: Course =  {
    courseId: courseId1,
    courseName: courseName1,
    timeZone: timeZone1,
    creationTimestamp: date1.getTime(),
    deletionTimestamp: 0,
  };
  const timeZoneOffsets1: Record<string, number> = { GMT: 0 };

  const spyStatusMessageService: any = {
    showErrorMessage: jest.fn(),
    showSuccessMessageTemplate: jest.fn(),
  };
  const timezoneServiceStub: any = {
    getTzOffsets: jest.fn(() => timeZoneOffsets1),
  };
  const spyCourseService: any = {
    createCourse: jest.fn(() => of(course1)),
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [AddCourseFormComponent],
      imports: [
        HttpClientTestingModule,
        FormsModule,
        ReactiveFormsModule,
        RouterTestingModule,
        NgbModule,
        MatSnackBarModule,
      ],
      providers: [
        { provide: StatusMessageService, useValue: spyStatusMessageService },
        { provide: CourseService, useValue: spyCourseService },
        { provide: TimezoneService, useValue: timezoneServiceStub },
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(AddCourseFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    component.timezone = timeZone1;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when not enabled', () => {
    component.isEnabled = false;
    component.timezone = timeZone1;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should call showErrorMessage when courseId is blank', () => {
    component.newCourseId = '';
    component.onSubmit();
    fixture.detectChanges();
    expect(spyStatusMessageService.showErrorMessage).toHaveBeenCalled();
  });

  it('should hold added course with valid details', () => {
    component.newCourseId = courseId1;
    component.newCourseName = courseName1;
    component.onSubmit();
    fixture.detectChanges();
    expect(spyStatusMessageService.showSuccessMessageTemplate).toHaveBeenCalled();
    expect(component.course).toEqual(course1);
  });
});
