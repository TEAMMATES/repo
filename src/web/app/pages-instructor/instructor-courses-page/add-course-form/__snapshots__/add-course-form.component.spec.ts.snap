// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCourseFormComponent should snap when not enabled 1`] = `
<tm-add-course-form
  closeCourseFormEvent={[Function EventEmitter]}
  courseAdded={[Function EventEmitter]}
  courseService={[Function Object]}
  isEnabled="false"
  newCourseId=""
  newCourseMessageTemplate={[Function TemplateRef_]}
  newCourseName=""
  statusMessageService={[Function Object]}
  timezone={[Function String]}
  timezoneService={[Function Object]}
  timezones={[Function Array]}
>
  <div
    class="card border-primary fill-plain"
  >
    <div
      class="close-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
      </button>
    </div>
    <div
      class="card-body"
    >
      <div
        class="form-group row"
      >
        <label
          class="col-sm-3 control-label bold-label text-right"
        >
          Course ID:
        </label>
        <div
          class="col-sm-3"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            maxlength="40"
            ng-reflect-maxlength="40"
            ng-reflect-model=""
            ng-reflect-ngb-tooltip="Enter the identifier of the co"
            ngbtooltip="Enter the identifier of the course, e.g.CS3215-2013Semester1."
            placeholder="e.g. CS3215-2013Semester1"
            type="text"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-sm-3 control-label bold-label text-right"
        >
          Course Name:
        </label>
        <div
          class="col-sm-9"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            maxlength="64"
            ng-reflect-maxlength="64"
            ng-reflect-model=""
            ng-reflect-ngb-tooltip="Enter the name of the course, "
            ngbtooltip="Enter the name of the course, e.g. Software Engineering."
            placeholder="e.g. Software Engineering"
            type="text"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-sm-3 control-label bold-label text-right"
        >
          Time Zone:
        </label>
        <div
          class="col-sm-9"
        >
          <div
            class="input-group"
          >
            <select
              class="form-control ng-untouched ng-pristine ng-valid"
              ng-reflect-model="UTC"
              ng-reflect-ngb-tooltip="The time zone for the course. "
              ngbtooltip="The time zone for the course. You should not
                  need to change this as it is auto-detected based on your device settings. TEAMMATES automatically adjusts
                  to match the current time offset in your area, including clock changes due to daylight saving time."
            >
              
              <option
                ng-reflect-value="GMT"
                value="GMT"
              >
                GMT
              </option>
            </select>
            <span
              class="input-group-btn"
            >
              <button
                class="btn btn-primary"
              >
                Auto Detect
              </button>
            </span>
          </div>
        </div>
      </div>
      <div
        class="form-group row"
      >
        <div
          class="offset-sm-3 col-sm-9"
        >
          <button
            class="btn btn-primary"
          >
            Add Course
          </button>
        </div>
      </div>
    </div>
  </div>
</tm-add-course-form>
`;

exports[`AddCourseFormComponent should snap with default fields 1`] = `
<tm-add-course-form
  closeCourseFormEvent={[Function EventEmitter]}
  courseAdded={[Function EventEmitter]}
  courseService={[Function Object]}
  isEnabled={[Function Boolean]}
  newCourseId=""
  newCourseMessageTemplate={[Function TemplateRef_]}
  newCourseName=""
  statusMessageService={[Function Object]}
  timezone={[Function String]}
  timezoneService={[Function Object]}
  timezones={[Function Array]}
>
  <div
    class="card border-primary fill-plain"
  >
    <div
      class="close-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
      </button>
    </div>
    <div
      class="card-body"
    >
      <div
        class="form-group row"
      >
        <label
          class="col-sm-3 control-label bold-label text-right"
        >
          Course ID:
        </label>
        <div
          class="col-sm-3"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            maxlength="40"
            ng-reflect-maxlength="40"
            ng-reflect-model=""
            ng-reflect-ngb-tooltip="Enter the identifier of the co"
            ngbtooltip="Enter the identifier of the course, e.g.CS3215-2013Semester1."
            placeholder="e.g. CS3215-2013Semester1"
            type="text"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-sm-3 control-label bold-label text-right"
        >
          Course Name:
        </label>
        <div
          class="col-sm-9"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            maxlength="64"
            ng-reflect-maxlength="64"
            ng-reflect-model=""
            ng-reflect-ngb-tooltip="Enter the name of the course, "
            ngbtooltip="Enter the name of the course, e.g. Software Engineering."
            placeholder="e.g. Software Engineering"
            type="text"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-sm-3 control-label bold-label text-right"
        >
          Time Zone:
        </label>
        <div
          class="col-sm-9"
        >
          <div
            class="input-group"
          >
            <select
              class="form-control ng-untouched ng-pristine ng-valid"
              ng-reflect-model="UTC"
              ng-reflect-ngb-tooltip="The time zone for the course. "
              ngbtooltip="The time zone for the course. You should not
                  need to change this as it is auto-detected based on your device settings. TEAMMATES automatically adjusts
                  to match the current time offset in your area, including clock changes due to daylight saving time."
            >
              
              <option
                ng-reflect-value="GMT"
                value="GMT"
              >
                GMT
              </option>
            </select>
            <span
              class="input-group-btn"
            >
              <button
                class="btn btn-primary"
              >
                Auto Detect
              </button>
            </span>
          </div>
        </div>
      </div>
      <div
        class="form-group row"
      >
        <div
          class="offset-sm-3 col-sm-9"
        >
          <button
            class="btn btn-primary"
          >
            Add Course
          </button>
        </div>
      </div>
    </div>
  </div>
</tm-add-course-form>
`;
